third_party/amd/lib/TritonAMDGPUTransforms/ConvertToBufferOps.cpp

```diff
@@ -32,35 +32,83 @@ namespace ttg = mlir::triton::gpu;
 namespace tt = mlir::triton;
 
 namespace {
-bool verifyNonNegativeByAssumption(Value expr,
-                                   const DenseSet<Value> &assumptions) {
+template <typename F>
+bool verifyNonSmallerByAssumption(Value expr,
+                                  const DenseSet<Value> &assumptions,
+                                  F matchesOther) {
   for (Value assume : assumptions) {
-    LDBG("Assumption:" << assume);
     if (auto cmpOp = assume.getDefiningOp<arith::CmpIOp>()) {
-      bool isGreaterThan = (cmpOp.getPredicate() == arith::CmpIPredicate::sge ||
-                            cmpOp.getPredicate() == arith::CmpIPredicate::sgt);
-      APInt cst;
-      if (isGreaterThan && (cmpOp.getLhs() == expr) &&
-          matchPattern(cmpOp.getRhs(), m_ConstantInt(&cst))) {
-        return cst.isNonNegative();
+      switch (cmpOp.getPredicate()) {
+      case arith::CmpIPredicate::eq:
+      case arith::CmpIPredicate::sge:
+      case arith::CmpIPredicate::sgt:
+      case arith::CmpIPredicate::uge:
+      case arith::CmpIPredicate::ugt: {
```

Comment:
I don't think the logic for `verifyNonNegativeByAssumption` is correct. You can't really tell if a value is negative based on an `uge` compare

Reply:
Yeah, in hindsight I agree this isn't quite necessarily right. E.g. here we'd be checking for `cmp uge %5, 0`, which is vacuously true (and would hopefully get elided at some point?)